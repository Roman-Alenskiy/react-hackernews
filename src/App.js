import React, { Component } from 'react'
import axios from 'axios'
import PropTypes from 'prop-types'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import './App.scss'

const DEFAULT_QUERY = 'react'
const DEFAULT_HPP = '50'

const PATH_BASE = 'https://hn.algolia.com/api/v1'
const PATH_SEARCH = '/search'
const PARAM_SEARCH = 'query='
const PARAM_PAGE = 'page='
const PARAM_HPP = 'hitsPerPage='

class App extends Component {
  _isMounted = false

  constructor(props) {
    super(props)

    this.state = {
      results: null,
      isLoading: false,
      searchKey: '',
      searchTerm: DEFAULT_QUERY,
      error: null,
    }

    this.needToSearchTopStories = this.needToSearchTopStories.bind(this)
    this.setSearchTopStories = this.setSearchTopStories.bind(this)
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this)
    this.onSearchChange = this.onSearchChange.bind(this)
    this.onSearchSubmit = this.onSearchSubmit.bind(this)
    this.onDismiss = this.onDismiss.bind(this)
  }

  needToSearchTopStories(searchTerm) {
    return !this.state.results[searchTerm]
  }

  setSearchTopStories(result) {
    const { hits, page } = result
    const { searchKey, results } = this.state

    const oldHits = results && results[searchKey]
      ? results[searchKey].hits
      : []
    
    const updatedHits = [...oldHits, ...hits]

    this.setState({ 
      results: {
        ...results,
        [searchKey]: { hits: updatedHits, page }
      },
      isLoading: false,
    })
  }

  onDismiss(id) {
    const { searchKey, results } = this.state
    const { hits, page } = results[searchKey]

    const isNotId = (item) => item.objectID !== id
    const updatedHits = hits.filter(isNotId)
    
    this.setState({ 
      results: {
        ...results, 
        [searchKey]: { hits: updatedHits, page }
      }
    })
  }

  onSearchChange(event) {
    this.setState({ searchTerm: event.target.value})
  }

  onSearchSubmit(event) {
    event.preventDefault()
    const { searchTerm } = this.state

    if (this.needToSearchTopStories(searchTerm)) {
      this.fetchSearchTopStories(searchTerm)
      this.setState({ isLoading: true })
    }

    this.setState({ 
      searchKey: searchTerm,
    })
  }

  onSearchMore(searchKey, page) {
    return () => {
      this.setState({ isLoading: true })

      this.fetchSearchTopStories(searchKey, page)
    }
  }

  componentDidMount() {
    this._isMounted = true
    const { searchTerm } = this.state
    this.setState({ 
      searchKey: searchTerm,
      isLoading: true 
    })
    this.fetchSearchTopStories(searchTerm)
  }

  componentWillUnmount() {
    this._isMounted = false
  }

  fetchSearchTopStories(searchTerm, page = 0) {

    axios(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
      .then(result => this._isMounted && this.setSearchTopStories(result.data))
      .catch(error => this._isMounted && this.setState({ error, isLoading: false }))
  }

  render() {
    const {
      results, isLoading, searchTerm, 
      searchKey, error
    } = this.state
    const page = (results && results[searchKey] && results[searchKey].page) || 0
    const list = (results && results[searchKey] && results[searchKey].hits) || []
    
    return (
      <div className="page">
        <div className="interactions">
          <Search
            value={searchTerm}
            onChange={this.onSearchChange}
            onSubmit={this.onSearchSubmit}
          >
            <p>
              Search
            </p>
          </Search>
        </div>
        <TableWithError
          error={error} 
          list={list}
          onDismiss={this.onDismiss}
        />
        <div className="interactions">
          <ButtonWithLoading
            className="more-hits-btn"
            onClick={this.onSearchMore(searchKey, page + 1)}
            isLoading={isLoading}
          >
            More
          </ButtonWithLoading>
        </div>
      </div>
    )
  }
}

class Search extends Component {
  constructor(props) {
    super(props)
    this.searchInput = React.createRef()
  }

  componentDidMount() {
    if (this.searchInput.current) this.searchInput.current.focus()
  }

  render() {
    const { value, onChange, onSubmit, children } = this.props

    return (
      <form onSubmit={onSubmit}>
        <input
          type="text"
          value={value}
          onChange={onChange}
          ref={this.searchInput}
        />
        <button type="submit">
          {children}
        </button>
      </form>
    )
  }
}

Search.propTypes = {
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
}

const Table = ({ list, onDismiss, children }) => {
  const largeColumn = {
    width: '40%',
  }
  const midColumn = {
    width: '30%',
  }
  const smallColumn = {
    width: '10%',
  } 

  return (
    <div className="table">
      {list.map((item) => {
        return (
          <div key={item.objectID} className="table-row">
            <span style={largeColumn}><a href={item.url} target="_blank" rel="noopener noreferrer">{item.title}</a></span>
            <span style={midColumn}>{item.author}</span>
            <span style={smallColumn}>{item.num_comments}</span>
            <span style={smallColumn}>{item.points}</span>
            <span style={smallColumn}>
              <Button
                className="button-inline"
                onClick={() => onDismiss(item.objectID)}
              >
                Dismiss
              </Button>
            </span>
          </div>
        )
      })}
      {children}
    </div>
  )
}

Table.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      objectID: PropTypes.string.isRequired,
      author: PropTypes.string,
      url: PropTypes.string,
      num_comments: PropTypes.number,
      points: PropTypes.number
    })
  ).isRequired,
  onDismiss: PropTypes.func.isRequired,
}

const Button = ({ onClick, className, children }) => (
  <button
    onClick={onClick}
    className={className}
    type="button"
  >
    {children}
  </button>
)

Button.defaultProps = {
  className: '',
}

Button.propTypes = {
  onClick: PropTypes.func.isRequired,
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
}

const LoadingIndicator = () => (
  <FontAwesomeIcon icon="spinner" className="loading" pulse/>
)

const withLoading = (Component) => (
  ({ isLoading, ...rest }) => ( 
    isLoading
    ? <LoadingIndicator />
    : <Component {...rest} />
  )
)

const withError = (Component) => (
  ({ error, ...rest }) => (
    error
    ? <div>
        <p>Something went wrong :(</p>
      </div>
    : <Component {...rest} />
  )
)

const ButtonWithLoading = withLoading(Button)
const TableWithError = withError(Table)

export default App

export {
  Button,
  Search,
  Table,
}