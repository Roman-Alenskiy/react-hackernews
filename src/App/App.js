import React, { Component } from 'react'
import axios from 'axios'
import './main.scss'

import { Search } from '../Search'
import { Table } from '../Table'
import { Button } from '../Button'
import { LoadingIndicator } from '../LoadingIndicator'

const DEFAULT_QUERY = 'react'
const DEFAULT_HPP = '50'

const PATH_BASE = 'https://hn.algolia.com/api/v1'
const PATH_SEARCH = '/search'
const PARAM_SEARCH = 'query='
const PARAM_PAGE = 'page='
const PARAM_HPP = 'hitsPerPage='

const updateSearchTopStoriesState = (hits, page) => (prevState) => {
  const { searchKey, results } = prevState

  const oldHits = results && results[searchKey]
    ? results[searchKey].hits
    : []

  const updatedHits = [...oldHits, ...hits]

  return {
    results: {
      ...results,
      [searchKey]: { hits: updatedHits, page }
    },
    isLoading: false,
  }
}

class App extends Component {
  _isMounted = false

  constructor(props) {
    super(props)

    this.state = {
      results: null,
      isLoading: false,
      searchKey: '',
      searchTerm: DEFAULT_QUERY,
      error: null,
    }

    this.needToSearchTopStories = this.needToSearchTopStories.bind(this)
    this.setSearchTopStories = this.setSearchTopStories.bind(this)
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this)
    this.onSearchChange = this.onSearchChange.bind(this)
    this.onSearchSubmit = this.onSearchSubmit.bind(this)
    this.onDismiss = this.onDismiss.bind(this)
  }

  needToSearchTopStories(searchTerm) {
    return !this.state.results[searchTerm]
  }

  setSearchTopStories(result) {
    const { hits, page } = result
    this.setState(updateSearchTopStoriesState(hits, page))
  }

  onDismiss(id) {
    const { searchKey, results } = this.state
    const { hits, page } = results[searchKey]

    const isNotId = (item) => item.objectID !== id
    const updatedHits = hits.filter(isNotId)
    
    this.setState({ 
      results: {
        ...results, 
        [searchKey]: { hits: updatedHits, page }
      }
    })
  }

  onSearchChange(event) {
    this.setState({ searchTerm: event.target.value})
  }

  onSearchSubmit(event) {
    event.preventDefault()
    const { searchTerm } = this.state

    if (this.needToSearchTopStories(searchTerm)) {
      this.fetchSearchTopStories(searchTerm)
      this.setState({ isLoading: true })
    }

    this.setState({ 
      searchKey: searchTerm,
    })
  }

  onSearchMore(searchKey, page) {
    return () => {
      this.setState({ isLoading: true })

      this.fetchSearchTopStories(searchKey, page)
    }
  }

  componentDidMount() {
    this._isMounted = true
    const { searchTerm } = this.state
    this.setState({ 
      searchKey: searchTerm,
      isLoading: true 
    })
    this.fetchSearchTopStories(searchTerm)
  }

  componentWillUnmount() {
    this._isMounted = false
  }

  fetchSearchTopStories(searchTerm, page = 0) {

    axios(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
      .then(result => this._isMounted && this.setSearchTopStories(result.data))
      .catch(error => this._isMounted && this.setState({ error, isLoading: false }))
  }

  render() {
    const {
      results, isLoading, searchTerm, 
      searchKey, error
    } = this.state
    const page = (results && results[searchKey] && results[searchKey].page) || 0
    const list = (results && results[searchKey] && results[searchKey].hits) || []
    
    return (
      <div className="page">
        <div className="interactions">
          <Search
            value={searchTerm}
            onChange={this.onSearchChange}
            onSubmit={this.onSearchSubmit}
          >
            <p>
              Search
            </p>
          </Search>
        </div>
        <TableWithError
          error={error} 
          list={list}
          onDismiss={this.onDismiss}
        />
        <div className="interactions">
          <ButtonWithLoading
            className="more-hits-btn"
            onClick={this.onSearchMore(searchKey, page + 1)}
            isLoading={isLoading}
          >
            More
          </ButtonWithLoading>
        </div>
      </div>
    )
  }
}

const withLoading = (Component) => (
  ({ isLoading, ...rest }) => ( 
    isLoading
    ? <LoadingIndicator />
    : <Component {...rest} />
  )
)

const withError = (Component) => (
  ({ error, ...rest }) => (
    error
    ? <div>
        <p>Something went wrong :(</p>
      </div>
    : <Component {...rest} />
  )
)

const ButtonWithLoading = withLoading(Button)
const TableWithError = withError(Table)

export default App

export {
  updateSearchTopStoriesState
}