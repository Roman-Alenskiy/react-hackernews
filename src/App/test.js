import React from 'react' 
import ReactDOM from 'react-dom' 
import renderer from 'react-test-renderer'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { App } from '.'
import { updateSearchTopStoriesState } from './App'
import { Search } from '../Search'

Enzyme.configure({ adapter: new Adapter() })

describe('App', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div') 
    ReactDOM.render(<App />, div) 
    ReactDOM.unmountComponentAtNode(div) 
  }) 

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <App />
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('passes value prop to Search', () => {
    const wrapper = shallow(<App />)
    let searchWrapper = wrapper.find(Search)
    
    expect(searchWrapper.props().value).toBe(wrapper.state().searchTerm)

    wrapper.setState({ searchTerm: 'redux' })

    searchWrapper = wrapper.find(Search)
    expect(searchWrapper.props().value).toBe('redux')
  })

})

describe('Helpers', () => {
  describe('updateSearchTopStoriesState()', () => {
    const prevState = {
      results: {
        query1: { 
          hits: [ {foo: 'foo'} ],
          page: 1
        }
      },
      isLoading: false,
      searchKey: 'query2'
    }

    const hits = [ {bar: 'bar'} ]
    const page = 1

    const updatedState = { 
      results: {
        query1: { 
          hits: [ {foo: 'foo'} ],
          page: 1
        },
        query2: {
          hits: [ {bar: 'bar'} ],
          page: 1
        }
      },
      isLoading: false,
    }

    it('successfully updates state', () => {
      expect(updateSearchTopStoriesState(hits, page)(prevState)).toEqual(updatedState)
    })
  })
})