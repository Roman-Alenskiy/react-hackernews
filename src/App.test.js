import React from 'react' 
import ReactDOM from 'react-dom' 
import renderer from 'react-test-renderer'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import App, { Search, Button, Table, updateSearchTopStoriesState } from './App' 

Enzyme.configure({ adapter: new Adapter() })

describe('App', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div') 
    ReactDOM.render(<App />, div) 
    ReactDOM.unmountComponentAtNode(div) 
  }) 

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <App />
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('passes value prop to Search', () => {
    const wrapper = shallow(<App />)
    let searchWrapper = wrapper.find(Search)
    
    expect(searchWrapper.props().value).toBe(wrapper.state().searchTerm)

    wrapper.setState({ searchTerm: 'redux' })

    searchWrapper = wrapper.find(Search)
    expect(searchWrapper.props().value).toBe('redux')
  })

})

describe('Search', () => {

  it('renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(
      <Search
        value=''
        onChange={() => {}}
        onSubmit={() => {}}
      >
        Search
      </Search>, div)
    ReactDOM.unmountComponentAtNode(div)
  })

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <Search
        value=''
        onChange={() => {}}
        onSubmit={() => {}}
      >
        Search
      </Search>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

})

describe('Button', () => {

  it('renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(<Button onClick={() => {}}>More</Button>, div)
    ReactDOM.unmountComponentAtNode(div)
  })

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <Button onClick={() => {}}>Give</Button>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

})

describe('Table', () => {

  const props = {
    list: [
      { title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y' },
      { title: '2', author: '2', num_comments: 1, points: 2, objectID: 'z' },
    ],
    onDismiss: () => {},
    sortKey: 'NONE',
    onSort: () => {},
    isSortReverse: false,
  }

  it('renders without crashing', () => {
    const div = document.createElement('div') 
    ReactDOM.render(<Table { ...props } />, div)
  }) 

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <Table { ...props } />
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('shows two items in list', () => {
    const element = shallow (
      <Table {...props} />
    )

    expect(element.find('.table-row').length).toBe(2)
  })

})

describe('Helpers', () => {
  describe('updateSearchTopStoriesState()', () => {
    const prevState = {
      results: {
        query1: { 
          hits: [ {foo: 'foo'} ],
          page: 1
        }
      },
      isLoading: false,
      searchKey: 'query2'
    }

    const hits = [ {bar: 'bar'} ]
    const page = 1

    const updatedState = { 
      results: {
        query1: { 
          hits: [ {foo: 'foo'} ],
          page: 1
        },
        query2: {
          hits: [ {bar: 'bar'} ],
          page: 1
        }
      },
      isLoading: false,
    }

    it('successfully updates state', () => {
      expect(updateSearchTopStoriesState(hits, page)(prevState)).toEqual(updatedState)
    })
  })
})