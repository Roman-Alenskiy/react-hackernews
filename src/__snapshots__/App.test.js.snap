// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        type="text"
        value="react"
      />
      <button
        type="submit"
      >
        <p>
          Search
        </p>
      </button>
    </form>
  </div>
  <div
    className="table"
  >
    <div
      className="table-header"
    >
      <span
        style={
          Object {
            "width": "40%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Title
          <i
            className="fa"
          />
        </button>
      </span>
      <span
        style={
          Object {
            "width": "30%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Author
          <i
            className="fa"
          />
        </button>
      </span>
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Comments
          <i
            className="fa"
          />
        </button>
      </span>
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Points
          <i
            className="fa"
          />
        </button>
      </span>
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      >
        Archive
      </span>
    </div>
  </div>
  <div
    className="interactions"
  >
    <i
      className="fa"
    />
  </div>
</div>
`;

exports[`Button has a valid snapshot 1`] = `
<button
  className=""
  onClick={[Function]}
  type="button"
>
  Give
</button>
`;

exports[`Search has a valid snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
  <button
    type="submit"
  >
    Search
  </button>
</form>
`;

exports[`Table has a valid snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-header"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Title
        <i
          className="fa"
        />
      </button>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Author
        <i
          className="fa"
        />
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Comments
        <i
          className="fa"
        />
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Points
        <i
          className="fa"
        />
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      Archive
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        rel="noopener noreferrer"
        target="_blank"
      >
        1
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        rel="noopener noreferrer"
        target="_blank"
      >
        2
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
</div>
`;
